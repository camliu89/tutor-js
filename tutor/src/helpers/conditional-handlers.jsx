import React from 'react';
import invariant from 'invariant';
import { extend, last, first } from 'lodash';
import { Redirect } from 'react-router-dom';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.isTeacher) {
      if (course.is_concept_coach) {
        return (
          <CCDashboard courseId={courseId} {...props} />
        );
      } else {
        return (
          <Redirect
            to={{
              pathname: Router.makePathname('viewTeacherDashboard', props.params),
              query: Router.currentQuery(),
            }} />
        );
      }
    } else {
      if (course.is_concept_coach) {
        return (
          <CCStudentRedirect courseId={courseId} />
        );
      } else {
        return <StudentDashboard {...props} />;
      }
    }
  };
  // eslint-disable-next-line react/prop-types
  const legacyReferenceBookRedirect = ({ params }) => {
    const parts = params.parts.split('/');
    const course = Courses.get(first(parts));
    invariant(course, `Did not find course for params '${params}'`);
    const chapterSection = parts.length > 1 ? last(parts) : 0;
    return (
      <Redirect
        to={{
          pathname: Router.makePathname(
            chapterSection ? 'viewReferenceBookSection' : 'viewReferenceBook',
            {
              ecosystemId: course.ecosystem_id,
              chapterSection,
            }),
          query: Router.currentQuery(),
        }}
      />
    );
  };

  return {
    dashboard() { return renderDashboard; },
    legacyReferenceBookRedirect() { return legacyReferenceBookRedirect; },
  };
};

export { getConditionalHandlers };
