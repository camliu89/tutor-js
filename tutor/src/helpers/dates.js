import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { compact, reduce, min, max, map } from 'lodash';
import TimeHelper from './time';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}

export function findEarliest(dateThings) {
  return min(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}

export function findLatest(dateThings) {
  return max(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}


export function dateWithUnchangedTime(date, t) {
  if (moment.isMoment(date)) {
    date = date.format(TimeHelper.ISO_DATE_FORMAT);
  }
  return TimeHelper.getDateOnly(date);
}
