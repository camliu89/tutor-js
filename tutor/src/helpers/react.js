import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
