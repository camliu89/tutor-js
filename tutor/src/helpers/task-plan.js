import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';
import moment from 'moment';
import Time from '../models/time';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },

  calendarParams(course) {
    let date;
    const dueAtRange = course
      .teacherTaskPlans.array
      .map(plan => plan.dateRanges.due.start)
      .sort();

    if (dueAtRange.length) {
      date = dueAtRange[0];
    } else {
      date = Time.now;
    }
    return {
      to: 'calendarByDate',
      params: {
        courseId: course.id,
        date: moment(date).format('YYYY-MM-DD'),
      },
    };
  },

  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },

};
