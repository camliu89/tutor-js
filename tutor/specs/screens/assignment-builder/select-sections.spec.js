import { C, FakeWindow } from '../../helpers';
import Factory, { FactoryBot } from '../../factories';
import SelectSections from '../../../src/screens/assignment-builder/select-sections';
import UX from '../../../src/screens/assignment-builder/ux';

jest.mock('../../../src/helpers/scroll-to');

describe('Select Topics', function() {
  let props;

  beforeEach(() => {
    const course = Factory.course();
    const plan = Factory.teacherTaskPlan({ course });
    const ux = new UX({ course, plan, windowImpl: new FakeWindow });
    ux.referenceBook.onApiRequestComplete({ data: [FactoryBot.create('Book')] });

    props = {
      ux,
      header: 'Section Chooser Header',
    };
  });

  it('matches snapshot', function() {
    expect.snapshot(<C><SelectSections {...props} /></C>).toMatchSnapshot();
  });

  it('scrolls into view on mount', () => {
    const ce = mount(<C><SelectSections {...props} /></C>);
    expect(props.ux.scroller.scrollToElement).toHaveBeenCalled();
    ce.unmount();
  });

});
