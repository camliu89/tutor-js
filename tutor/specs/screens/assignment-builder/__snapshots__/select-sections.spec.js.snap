// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Topics matches snapshot 1`] = `
.c5 {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

.c1 {
  max-width: 1200px;
  margin: auto;
}

.c6 {
  font-weight: 700;
  color: #424242;
}

.c7 {
  display: inline-block;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 0.5rem;
  cursor: pointer;
  font-size: 1.5rem;
  border-bottom: 1px solid #f1f1f1;
}

.c8 > * {
  margin-left: 1rem;
}

.c8 input {
  font-size: 1.7rem;
  margin-left: 1.2rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  font-size: 1.7rem;
  border-bottom: 1px solid #f1f1f1;
}

.c4 > * {
  margin-left: 1rem;
}

.c4 .chapter-title {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 .browse-the-book {
  font-size: 1.4rem;
  margin-right: 2rem;
}

.c2 {
  border: 1px solid #f1f1f1;
  border-radius: 4px;
}

.c0 {
  margin: 4rem auto;
}

<div
  className="openstax-debug-content"
>
  <div
    data-purpose="tour-conductor-wrapper"
  >
    <div>
      <div
        className="c0 reading-plan-select-sections"
        data-tour-region-id="add-reading-choose-sections"
      >
        <div
          className="c1 dialog default-dialog select-sections card"
        >
          <div
            className="card-header"
          >
            Section Chooser Header
            <button
              className="openstax-close-x close"
              onClick={[Function]}
            />
          </div>
          <div
            className="select-chapters"
          >
            <div
              className="c2"
            >
              <div
                className="c3 chapter"
                data-is-expanded={true}
              >
                <div
                  className="c4"
                  data-chapter-section={1}
                  onClick={[Function]}
                  role="button"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <button
                        className="ox-icon-square btn btn-plain"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-square fa-w-14 c5 ox-icon ox-icon-square"
                          color="#929292"
                          data-icon="square"
                          data-prefix="far"
                          focusable="false"
                          role="img"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </button>
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="1.0"
                    >
                      1
                    </span>
                     - 
                  </span>
                  <div
                    className="c7 chapter-title"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "Kinematics",
                      }
                    }
                  />
                  <a
                    aria-label="Browse course reference book"
                    className="browse-the-book"
                    href="/book/1/page/234"
                    onClick={[Function]}
                    target="_blank"
                  >
                    Browse this Chapter
                  </a>
                </div>
                <div
                  aria-expanded={null}
                  className="sections collapse show"
                >
                  <div
                    className="c8 section"
                    data-section-id="234"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="1.1"
                    >
                      1.1
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Kinematics in 1 dim",
                        }
                      }
                    />
                  </div>
                  <div
                    className="c8 section"
                    data-section-id="235"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="1.2"
                    >
                      1.2
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Kinematics in 2 dim",
                        }
                      }
                    />
                  </div>
                  <div
                    className="c8 section"
                    data-section-id="236"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="1.3"
                    >
                      1.3
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Kinematics in 3 dim",
                        }
                      }
                    />
                  </div>
                </div>
              </div>
              <div
                className="c3 chapter"
                data-is-expanded={false}
              >
                <div
                  className="c4"
                  data-chapter-section={2}
                  onClick={[Function]}
                  role="button"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <button
                        className="ox-icon-square btn btn-plain"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-square fa-w-14 c5 ox-icon ox-icon-square"
                          color="#929292"
                          data-icon="square"
                          data-prefix="far"
                          focusable="false"
                          role="img"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </button>
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="2.0"
                    >
                      2
                    </span>
                     - 
                  </span>
                  <div
                    className="c7 chapter-title"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "Mechanics",
                      }
                    }
                  />
                  <a
                    aria-label="Browse course reference book"
                    className="browse-the-book"
                    href="/book/1/page/237"
                    onClick={[Function]}
                    target="_blank"
                  >
                    Browse this Chapter
                  </a>
                </div>
                <div
                  aria-expanded={null}
                  className="sections collapse"
                >
                  <div
                    className="c8 section"
                    data-section-id="237"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="2.1"
                    >
                      2.1
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Mechanics in 1 dim",
                        }
                      }
                    />
                  </div>
                  <div
                    className="c8 section"
                    data-section-id="238"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="2.2"
                    >
                      2.2
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Mechanics in 2 dim",
                        }
                      }
                    />
                  </div>
                  <div
                    className="c8 section"
                    data-section-id="239"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="2.3"
                    >
                      2.3
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Mechanics in 3 dim",
                        }
                      }
                    />
                  </div>
                </div>
              </div>
              <div
                className="c3 chapter"
                data-is-expanded={false}
              >
                <div
                  className="c4"
                  data-chapter-section={3}
                  onClick={[Function]}
                  role="button"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <button
                        className="ox-icon-square btn btn-plain"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-square fa-w-14 c5 ox-icon ox-icon-square"
                          color="#929292"
                          data-icon="square"
                          data-prefix="far"
                          focusable="false"
                          role="img"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </button>
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="3.0"
                    >
                      3
                    </span>
                     - 
                  </span>
                  <div
                    className="c7 chapter-title"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "Thermodynamics",
                      }
                    }
                  />
                  <a
                    aria-label="Browse course reference book"
                    className="browse-the-book"
                    href="/book/1/page/240"
                    onClick={[Function]}
                    target="_blank"
                  >
                    Browse this Chapter
                  </a>
                </div>
                <div
                  aria-expanded={null}
                  className="sections collapse"
                >
                  <div
                    className="c8 section"
                    data-section-id="240"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="3.1"
                    >
                      3.1
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Thermodynamics in 1 dim",
                        }
                      }
                    />
                  </div>
                  <div
                    className="c8 section"
                    data-section-id="241"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="3.2"
                    >
                      3.2
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Thermodynamics in 2 dim",
                        }
                      }
                    />
                  </div>
                  <div
                    className="c8 section"
                    data-section-id="242"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="c6 chapter-section"
                      data-chapter-section="3.2"
                    >
                      3.2
                    </span>
                    <div
                      className="c7 section-title"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "Thermodynamics in 3 dim",
                        }
                      }
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="card-footer"
          >
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="visible-when-debugging"
    >
      <div
        className="tour-spy-info"
      >
        available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
      </div>
    </div>
  </div>
  <a
    aria-hidden="true"
    className="debug-toggle-link"
    href="#spy"
    onClick={[Function]}
    tabIndex={-1}
  >
    π
  </a>
</div>
`;
